server/src/main/java/org/candlepin/resource/validation/DTOValidator.java

```diff
@@ -0,0 +1,239 @@
+     * }
+     * </pre>
+     *
+     * @param getters a variable amount of method references to getters that return Collections of items
+     * @throws IllegalArgumentException when any of the collections returned from the specified getters
+     * contains at least one null element
+     */
+    @SafeVarargs
+    public final void validateCollectionElementsNotNull(Supplier<? extends Collection<?>> getter1,
+        Supplier<? extends Collection<?>> getter2, Supplier<? extends Collection<?>> getter3,
+        Supplier<? extends Collection<?>>... getters) {
+
+        validateCollectionElementsNotNull(getter1, getter2, getter3);
+
+        for (Supplier<? extends Collection<?>> getter : getters) {
+            Collection<?> collection = getter.get();
+            if (isCollectionInvalid(collection)) {
+                throw new IllegalArgumentException("collection contains null elements");
+            }
+        }
+    }
+
+    /*
+     * ======================================================================================================
+     * The following methods are overloaded versions of validateCollectionElementsNotNull for one, two and
+     * three getters respectively, as a performance optimization to avoid the implicit creation of an array
+     * object that happens during the vararg method call. There are currently no DTOs that contain more than
+     * three collections that require null element validation, so this optimisation would cover 100% of cases
+     * for now, while allowing us to validate more than 3 collections at a time in the future.
+     * ======================================================================================================
+     */
+
+    /**
+     * Works the same as {@link #validateCollectionElementsNotNull(Supplier, Supplier, Supplier, Supplier...)}
+     */
+    public void validateCollectionElementsNotNull(Supplier<? extends Collection<?>> getter) {
+        Collection<?> collection = getter.get();
+        if (isCollectionInvalid(collection)) {
+            throw new IllegalArgumentException("collection contains null elements");
+        }
+    }
+
+    /**
+     * Works the same as {@link #validateCollectionElementsNotNull(Supplier, Supplier, Supplier, Supplier...)}
+     */
+    public void validateCollectionElementsNotNull(Supplier<? extends Collection<?>> getter1,
+        Supplier<? extends Collection<?>> getter2) {
+        Collection<?> collection1 = getter1.get();
+        Collection<?> collection2 = getter2.get();
+        if (isCollectionInvalid(collection1) ||
+            isCollectionInvalid(collection2)) {
+            throw new IllegalArgumentException("collection contains null elements");
+        }
+    }
+
+    /**
+     * Works the same as {@link #validateCollectionElementsNotNull(Supplier, Supplier, Supplier, Supplier...)}
+     */
+    public void validateCollectionElementsNotNull(Supplier<? extends Collection<?>> getter1,
```

Comment:
I'm mostly alright with the microoptimization here around the avoidance of the implicit array creation, however, I feel these should be structured with code minimization in mind -- especially since we're offloading the work to a worker method to begin with.

Anyway, since we have a private method which validates a collection, and the single-param version does nothing more than unpack the collection from its supplier, it feels like the cleanest way to do all of this is to declare the single-param version as the base, and then have the various variadic overloads call it with each supplier. For example:

```
    public void validateCollectionElementsNotNull(Supplier<? extends Collection<?>> getter) {
        Collection<?> collection = getter.get();
        if (collection == null || collection.stream().anyMatch(Objects::isNull)) {
            throw new IllegalArgumentException("collection contains null elements");
        }
    }

    public void validateCollectionElementsNotNull(Supplier<? extends Collection<?>> getter1,
        Supplier<? extends Collection<?>> getter2) {

        this.validateCollectionElementsNonNull(getter1);
        this.validateCollectionElementsNonNull(getter2);
    }

    // and so on
```

Reply:
Good idea

Reply:
Done